File: bench/src/go/dispatch.go
package main
import "os"
type Shape interface{ area() int64 }
type Circle struct{ r int64 }
type Square struct{ s int64 }
func (c Circle) area() int64 { return c.r*c.r }
func (q Square) area() int64 { return q.s*q.s }
func measure(sh Shape) int64 { return sh.area() }
func main(){ var c Shape = Circle{3}; var q Shape = Square{4}; var sum uint64
  for i:=uint64(0);i<8000000;i++ { if i%2==0 { sum+=uint64(measure(c)) } else { sum+=uint64(measure(q)) } }
  os.Exit(int(sum%256)) }

File: bench/src/lua/dispatch.lua
-- dispatch — dynamic dispatch over a heterogeneous table of shapes built at runtime.
local function make_circle(r) return { area = function() return r*r end } end
local function make_square(s) return { area = function() return s*s end } end
local N = 512
local shapes = {}
for j=0,N-1 do if j%2==0 then shapes[j] = make_circle(3) else shapes[j] = make_square(4) end end
local sum = 0
for i=0,7999999 do sum = sum + shapes[i%N].area() end
os.exit(sum % 256)
